core/sequencer/server.go

```diff
@@ -131,55 +145,68 @@ func (s *Server) RunBatch(ctx context.Context, in *spb.RunBatchRequest) (*empty.
 	if err := proto.Unmarshal(latestMapRoot.Metadata, &lastMeta); err != nil {
 		return nil, err
 	}
-	highs, err := s.logs.HighWatermarks(ctx, in.DomainId)
+
+	startWatermarks := make(Watermarks)
+	for logID, source := range lastMeta.Sources {
+		startWatermarks[logID] = source.HighestWatermark
+	}
+	count, highs, err := s.HighWatermarks(ctx, in.DomainId, startWatermarks, in.MaxBatch)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "HighWatermark(): %v", err)
+		return nil, status.Errorf(codes.Internal, "HighWatermarks(): %v", err)
 	}
-
-	// TODO(#1057): If time since last map revision > max timeout, run batch.
-	// TODO(#1047): If time since oldest log item > max latency has elapsed, run batch.
-	// TODO(#1056): If count items > max_batch, run batch.
-
-	// Count items to be processed.  Unfortunately, this means we will be
-	// reading the items to be processed twice.  Once, here in RunBatch and
-	// again in CreateEpoch.
-	sources := make(map[int64]*spb.MapMetadata_SourceSlice)
-	for sliceID, high := range highs {
-		sources[sliceID] = &spb.MapMetadata_SourceSlice{
-			LowestWatermark:  lastMeta.Sources[sliceID].GetHighestWatermark(),
+	sources := make(SourcesEntry)
+	for logID, high := range highs {
+		sources[logID] = &spb.MapMetadata_SourceSlice{
+			LowestWatermark:  lastMeta.Sources[logID].GetHighestWatermark(),
```

Comment:
Is it the same as `startWatermarks[logID]`? Could be shortened.

Reply:
Yes, good catch.